feat(component): SfRange implementation

AdamPawlinski opened the pull request:
# Related issue
Closes #852 

# Scope of work
New component added to molecules: SfRange.
It consists of: 
- two SfInput components for two thumbs (min and max), 
- label for whole component,
- two labels for both thumbs with current values.

Props added: 
- min
- max
- step
- label
- labelMin - label not visible only for a11y purposes
- labelMax - label not visible only for a11y purposes

Component needs styling (waits for final design) and in my opinion, adding possibility to change for vertical (modifier) and one thumb option (prop).     

# Screenshots of visual changes


# Checklist

- [x] No commented blocks of code left
- [x] Run tests and docs
- [x] Self code-reviewed

pspaczek commented:
@AdamPawlinski can you add API for this component to PR description? :)

AdamPawlinski commented:
@pspaczek yes obviosly but I still haven't prepared it, because these component is not ready as multi-thumb slider. I would like to prepare it first and then write the API down in PR.

AdamPawlinski commented:
I fixed some uislider issues, but there is still problem with implementation of change orientation prop. I didn't figure it out how to make it work.

Michal-Dziedzinski commented:
@AdamPawlinski grat job! Did you have the idea that the outline should be visible when using mouse?

![image](https://user-images.githubusercontent.com/32803679/87656320-a0ae0f80-c759-11ea-99b0-a81d03f55607.png)

Michal-Dziedzinski commented on packages/vue/src/components/molecules/SfRange/SfRange.vue:
```diff
@@ -0,0 +1,169 @@
+          return this.updateOptions({
+            range: {
+              min: value,
+              max: this.max,
```
Shouldn't a condition be checked here as in `mounted`?

AdamPawlinski replied:
Yes you are right :)

Michal-Dziedzinski commented on packages/vue/src/components/molecules/SfRange/SfRange.vue:
```diff
@@ -0,0 +1,169 @@
+      handler(value) {
+        return this.updateOptions({
+          range: {
+            min: this.min,
```
Same question as above?

AdamPawlinski replied:
like above :)

AdamPawlinski commented on packages/shared/styles/components/molecules/SfRange.scss:
```diff
@@ -0,0 +1,57 @@
+  );
+  border: none;
+  box-shadow: none;
+  .noUi-handle {
```
Yes it's obvious I should have thought about it. Thanks.

AdamPawlinski commented on packages/vue/src/components/molecules/SfRange/SfRange.vue:
```diff
@@ -0,0 +1,169 @@
+      default: 0,
+    },
+    valueMax: {
+      type: [Number, String],
```
Probably it left unchanged after refactoring I made, but should be changed.

AdamPawlinski commented on packages/vue/src/components/molecules/SfRange/SfRange.vue:
```diff
@@ -0,0 +1,169 @@
+  mounted() {
+    this.config = {
+      range: {
+        min: this.min,
```
Weird. I'll fix that.

justyna-13 commented on packages/vue/src/components/molecules/SfRange/SfRange.stories.js:
```diff
@@ -0,0 +1,109 @@
+      disabled: {
+        default: boolean("disabled", false, "Props"),
+      },
+      orientation: {
```
what do you think about changing this name on horizontalOrientation? It's boolean and I think this name is more readable :)

AdamPawlinski replied:
Yes you are right. It will be more meaningful.

justyna-13 commented on packages/vue/src/components/molecules/SfRange/SfRange.stories.js:
```diff
@@ -0,0 +1,109 @@
+      orientation: {
+        default: boolean("orientation", true, "Props"),
+      },
+      direction: {
```
same here :) I'd change this name on ltrDirection

AdamPawlinski replied:
Like above

AdamPawlinski commented:
CR changes applied. Lib for formatting values and prop to configure format added. There should be solution for showing outline on keyup event added for a11y reasons, but since focus directive doesn't work I did't figure out how to do this.

AdamPawlinski commented on packages/vue/src/components/molecules/SfRange/SfRange.spec.js:
```diff
@@ -0,0 +1,9 @@
+describe("SfRange.vue", () => {
+  it("renders a component", () => {
+    const component = shallowMount(SfRange);
+    expect(component.contains(".sf-range")).toBe(true);
```
Thanks, I'll change that.

justyna-13 commented on .gitignore:
```diff
@@ -1,6 +1,7 @@
 package-lock.json
 .DS_Store
 node_modules
+vsf-test
```
this shouldn't be here:)

justyna-13 reviewed (changes requested):
Looks great the only thing is a11y that needs to be improved :) tab clicking doesn't select handlers

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/DivanteLtd/storefront-ui?pullRequest=1230) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/DivanteLtd/storefront-ui?pullRequest=1230) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: justyna-13<br/>:white_check_mark: AdamPawlinski<br/>:x: Przemek Spaczek<br/><hr/>**Przemek Spaczek** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/DivanteLtd/storefront-ui?pullRequest=1230) it.</sub>

Fifciu commented on packages/vue/src/components/molecules/SfRange/SfRange.vue:
```diff
@@ -0,0 +1,231 @@
+        return this.updateRangeOptions({
+          range: {
+            min: this.min,
+            max: value <= this.min ? this.min + 1 : value,
```
Same as above

AdamPawlinski replied:
like above

Fifciu commented on packages/vue/src/components/molecules/SfRange/SfRange.vue:
```diff
@@ -0,0 +1,231 @@
+    },
+    horizontalOrientation: {
+      handler(value) {
+        value ? (value = "horizontal") : (value = "vertical");
```
Do you need it? Couldn't we below just use:
```js
orientation: value ? 'horizontal' : 'vertical',
```

AdamPawlinski replied:
it changes boolean prop into value = horizontal or vertical, maybe I change the value into orientationValue to be clearer: 
```
handler(value) {
				let orientationValue = "horizontal";
        value ? orientationValue = "horizontal" : orientationValue = "vertical";
        return this.resetAndChangeOption({
          orientation: orientationValue,
        });
      },
```

Fifciu commented on packages/vue/src/components/molecules/SfRange/SfRange.vue:
```diff
@@ -0,0 +1,231 @@
+    },
+    ltrDirection: {
+      handler(value) {
+        value ? (value = "ltr") : (value = "rtl");
```
Same as above

AdamPawlinski replied:
like above

Fifciu commented on packages/vue/src/components/molecules/SfRange/SfRange.vue:
```diff
@@ -0,0 +1,231 @@
+    },
+    resetAndChangeOption(changedValue) {
+      if (this.$refs.range) {
+        this.$refs.range.noUiSlider.destroy();
```
I wonder if we maybe should also destroy it in `beforeDestroy` to prevent memory leak

AdamPawlinski replied:
Ok I add it to to beforeDestroy method.

AdamPawlinski commented on packages/vue/src/components/molecules/SfRange/SfRange.vue:
```diff
@@ -0,0 +1,231 @@
+    tooltips: {
+      handler(value) {
+        value
+          ? (value =
```
Yes, you are right, I'll change this.

AdamPawlinski commented on packages/vue/src/components/molecules/SfRange/SfRange.vue:
```diff
@@ -0,0 +1,231 @@
+        if (this.$refs.range) {
+          return this.updateRangeOptions({
+            range: {
+              min: value >= this.max ? this.max - 1 : value,
```
When I set min and max to be equal to 0 error appeared from nouislider lib: 
![image](https://user-images.githubusercontent.com/32042425/99389829-e8939d80-28d7-11eb-86a9-7b52f6181038.png)
 cannot set this values to equal. Nevertheless, I wonder if I should change this watchers  and use step value to differantiate max from min, wdyt?
``` min: value >= this.max ? this.max - this.step : value ```

Fifciu commented on packages/vue/src/components/molecules/SfRange/SfRange.vue:
```diff
@@ -0,0 +1,226 @@
+  watch: {
+    min: {
+      handler(value) {
+        if (this.$refs.range) {
```
You already have this condition inside `updateRangeOptions` method

AdamPawlinski replied:
thanks, I changed this

Fifciu commented on packages/vue/src/components/molecules/SfRange/SfRange.vue:
```diff
@@ -0,0 +1,226 @@
+            this.slidersInitialValues.length === 2
+              ? [
+                  wNumb(this.formatTooltipsValues),
+                  wNumb(this.formatTooltipsValues),
```
Twice same?
Maybe it would be worth to set some const:
```ts
const wnumbed = wNumb(this.formatTooltipsValues);

// and then
[ wnumbed, wnumbed ]
// and
[ wnumbed ]
```

AdamPawlinski replied:
it looks better, thanks :)

Fifciu commented on packages/vue/src/components/molecules/SfRange/SfRange.vue:
```diff
@@ -0,0 +1,226 @@
+            tooltips:
+              this.slidersInitialValues.length === 2
+                ? [
+                    wNumb(this.formatTooltipsValues),
```
Same here with the const

AdamPawlinski replied:
wNumb was removed from component

Fifciu commented on packages/vue/src/components/molecules/SfRange/SfRange.vue:
```diff
@@ -0,0 +1,226 @@
+        let orientationValue = "horizontal";
+        value
+          ? (orientationValue = "horizontal")
+          : (orientationValue = "vertical");
```
```js
let orientationValue = value ? 'horizontal' : 'vertical'
```

Fifciu commented on packages/vue/src/components/molecules/SfRange/SfRange.vue:
```diff
@@ -0,0 +1,226 @@
+    ltrDirection: {
+      handler(value) {
+        let directionValue = "ltr";
+        value ? (directionValue = "ltr") : (directionValue = "rtl");
```
let directionValue = value ? 'ltr' : 'rtl'

Fifciu commented on packages/vue/src/components/molecules/SfRange/SfRange.vue:
```diff
@@ -0,0 +1,226 @@
+    },
+  },
+  mounted() {
+    this.config = {
```
Maybe it should be a computed value?

Fifciu commented on packages/vue/src/components/molecules/SfRange/SfRange.vue:
```diff
@@ -0,0 +1,226 @@
+    this.noUiSliderInit();
+  },
+  beforeDestroy() {
+    this.$refs.range.noUiSlider.destroy();
```
if (this.$refs.range && this.$refs.range.noUiSlider)

filipsobol replied:
`this.$refs.range?.noUiSlider?.destroy();`

filipsobol commented on packages/vue/src/components/molecules/SfRange/SfRange.vue:
```diff
@@ -0,0 +1,226 @@
+      direction: this.ltrDirection ? "ltr" : "rtl",
+      orientation: this.horizontalOrientation ? "horizontal" : "vertical",
+      behaviour: "tap-drag",
+      tooltips: false,
```
```
tooltips: this.tooltips,
```

filipsobol commented:
Add margins / paddings at the bottom of the component if tooltips are enabled

AdamPawlinski commented:
I simplified component removing wNumb lib and use js native Intl.Number formatting and also added range to Form example to show such component in bigger context,

mayashavin commented on packages/shared/styles/components/molecules/SfRange.scss:
```diff
@@ -4,7 +4,7 @@
   position: relative;
   width: var(--range-width, 100%);
   height: var(--range-height, 7px);
```
7px? set it as `--range-height` default value please

AdamPawlinski replied:
ok I changed it.

mayashavin commented on packages/shared/styles/components/molecules/SfRange.scss:
```diff
@@ -4,7 +4,7 @@
   height: var(--range-height, 7px);
-  padding: var(--range-tooltips-padding, 0 var(--spacer-sm));
+  margin: var(--spacer-xl);
   @include background(--range-background, var(--c-light));
```
remove --range-background

AdamPawlinski replied:
Yes of course.

mayashavin commented on packages/vue/src/components/atoms/SfImage/SfImage.stories.js:
```diff
@@ -28,64 +28,10 @@ export default {
         category: "Props",
       },
     },
-    template: `<SfImage
```
why remove all these stories?

mayashavin commented on packages/vue/src/components/molecules/SfRange/SfRange.stories.js:
```diff
@@ -1,117 +1,136 @@
-            },
-            keyboardSupport: true,
+      defaultValue: {
+        start: [0,1],
```
make it flat instead of nested object of props

AdamPawlinski replied:
Do you mean to flatten it in the story, like did now. Or should it be changed in component props?

mayashavin reviewed (changes requested):
Please try to flatten out `props` instead of nested object inside object, such as `config`. That's not a good practice.

AdamPawlinski commented:
Branch pushed to original repo: https://github.com/vuestorefront/storefront-ui/tree/%23852-feat/sfrange-implement

Baroshem commented on packages/vue/src/components/molecules/SfRange/SfRange.vue:
```diff
@@ -0,0 +1,87 @@
+      handler(values) {
+        return this.$refs.range.noUiSlider.set(values);
+      },
+      immedite: true,
```
I don't think this will work ;) should be `immediate` https://vuejs.org/v2/api/#watch

AdamPawlinski replied:
thanks :)

Baroshem commented on packages/vue/src/components/molecules/SfRange/SfRange.vue:
```diff
@@ -0,0 +1,87 @@
+          this.$emit("change", values);
+        })
+    },
+    resetAndChangeOption(config) {
```
Method should always do one thing. Here it resets something and changes some option. Considering the fact that this method is used in only one place (line 51) I think it does not need to be encapsulated. In line 51 instead of just returning a method you should just destroy the noUiSlider instance and init new one.

AdamPawlinski replied:
Yes you are right thanks :)

Baroshem commented on packages/vue/src/components/molecules/SfRange/SfRange.vue:
```diff
@@ -0,0 +1,87 @@
+    },
+  },
+  watch: {
+    config: {
```
Do we expect config to change over time? I think it should be initialized once and used throughout the app. Correct me if I am wrong ;)

AdamPawlinski replied:
For norrmal use cases you are right, but I also consider what is going on in Storybook, when data can change. If I shouldn't do this I'll remove this part.

Baroshem commented on packages/vue/src/components/molecules/SfRange/SfRange.vue:
```diff
@@ -0,0 +1,87 @@
+    /*
+     * Sets the number of sliders (1 or 2) by adding initial values to the array
+     */
+    value: {
```
I don't think this is a good name for that prop. Maybe `numberOfSliders`?

AdamPawlinski replied:
I think the description for this property is wrong. It should be that this are starting values for sliders and when there is one number given only one slider appears.

Baroshem commented on packages/vue/src/components/molecules/SfRange/SfRange.vue:
```diff
@@ -0,0 +1,87 @@
+     * Sets the number of sliders (1 or 2) by adding initial values to the array
+     */
+    value: {
+      type: [Number, Array],
```
Also, how do you expect to handle the case when someone passes number as a prop? (`value="3"`). I think it would be better to just expect one type and handle it, also warn user that there is a wrong prop type passed.

AdamPawlinski replied:
It can be passed to the nouiSlder config as an array or a number, but If you think one type is better I remove that,

Baroshem commented on packages/vue/src/examples/form/Form.vue:
```diff
@@ -101,6 +101,12 @@
       />
+      <SfHeading title="Prices" level="5" />
+      <SfRange
+        v-model="range"
```
I think there was a change in the naming of the props/variables. Type and value of `range`  looks similar to `value` in SfRange. Is it the same? Also I don't think it is being used in SfRange

AdamPawlinski replied:
Here I'm passing array of values for setting sliders in SfRange. I think it's okay, but maybe I'm wrong.

AdamPawlinski commented on packages/shared/styles/components/molecules/SfRange.scss:
```diff
@@ -0,0 +1,84 @@
+  background-color: var(--c-light);
+  border: none;
+  box-shadow: none;
+  .noUi {
```
It's the class for styling elements from nouiSlider library needed to make this component work.

AdamPawlinski commented on packages/shared/styles/components/molecules/SfRange.scss:
```diff
@@ -0,0 +1,84 @@
+    }
+  }
+}
+.noUi-vertical {
```
Like above it's element from this lib.

justyna-13 reviewed (changes requested):
I think it looked a bit different before :) please take a look at it 
![Screenshot from 2021-06-07 12-32-40](https://user-images.githubusercontent.com/46591755/121002706-e5d19a80-c78c-11eb-9d8c-64aa2c66b816.png)

AdamPawlinski commented:
I created another PR for this created from hotfix-0.10.7 branch https://github.com/vuestorefront/storefront-ui/pull/1820
